core/src/main/java/org/elasticsearch/common/util/concurrent/QueueResizingEsThreadPoolExecutor.java

```diff
@@ -0,0 +1,210 @@
+        if (t == null) {
+            assert r instanceof TimedRunnable : "expected only TimedRunnables in queue";
+            final long taskNanos = ((TimedRunnable) r).getTotalNanos();
+            final long totalNanos = totalTaskNanos.addAndGet(taskNanos);
+            if (taskCount.incrementAndGet() == this.tasksPerWindow) {
+                final long endTimeNs = System.nanoTime();
+                final long totalRuntime = endTimeNs - this.startNs;
+                // Reset the start time for all tasks. Theoretically this needs to be volatile, since
+                // we are reading from a different thread when it is set, but it is protected by the
+                // `taskCount` memory barrier.
+                startNs = endTimeNs;
+
+                // Calculate the new desired queue size
+                try {
+                    int desiredQueueSize = calculateL(totalRuntime, tasksPerWindow, totalNanos);
+                    logger.debug("estimated optimal queue size for [{}] threadpool: [{}]", name, desiredQueueSize);
+
+                    final int oldCapacity = workQueue.capacity();
+
+                    // Adjust the queue size towards the desired capacity using an adjust of
+                    // QUEUE_ADJUSTMENT_AMOUNT (either up or down), keeping in mind the min and max
+                    // values the queue size can have.
+                    final int newCapacity = workQueue.adjustCapacity(desiredQueueSize,
+                            this.queueAdjustmentAmount, minQueueSize, maxQueueSize);
+                    if (oldCapacity != newCapacity) {
+                        logger.debug("adjusted [{}] queue by [{}], old capacity: [{}], new capacity: [{}]", name,
+                                newCapacity > oldCapacity ? this.queueAdjustmentAmount : -this.queueAdjustmentAmount,
+                                oldCapacity, newCapacity);
+                    }
+                } catch (ArithmeticException e) {
+                    // There was an integer overflow, so just log about it, rather than adjust the queue size
+                    logger.warn((Supplier<?>) () -> new ParameterizedMessage(
+                                    "failed to calculate optimal queue size for [{}] thread pool, " +
+                                    "total frame time [{}ns], tasks [{}], task execution time [{}ns]",
+                                    name, totalRuntime, tasksPerWindow, totalNanos),
+                            e);
+                } finally {
+                    // Finally, decrement the task count and time back to their starting values. We
+                    // do this at the end so there is no concurrent adjustments happening. We also
+                    // decrement them instead of resetting them back to zero, as
+                    int tasks = taskCount.addAndGet(-this.tasksPerWindow);
+                    assert tasks >= 0 : "tasks should never less than 0";
```

Comment:
Add the offending value? If this breaks, it's going to be in tests in concurrent code where it's possible it will not reproduce easily to any hint as to the problem will be welcome.